File: fvf4j/src/main/java/fvf4j/demo/domain/Transaction.java
package fvf4j.demo.domain;


import fvf4j.demo.domain.CategorizedTransaction.ExpenseCategory;

import java.math.BigDecimal;
import java.util.Objects;


public record Transaction(
        TransactionId transactionId,
        ClientId clientId,
        AccountId accountId,
        Amount amount,
        MerchantCategoryCode mcc
) {
    public Transaction {
        Objects.requireNonNull(transactionId, "Transaction ID cannot be null");
        Objects.requireNonNull(clientId, "Client ID cannot be null");
        Objects.requireNonNull(accountId, "Account ID cannot be null");
        Objects.requireNonNull(amount, "Amount cannot be null");
        Objects.requireNonNull(mcc, "Merchant category code cannot be null");
    }

    public CategorizedTransaction toCategorizedTransaction(ExpenseCategory expenseCategory) {
        return new CategorizedTransaction(
                null,
                transactionId(),
                clientId(),
                accountId(),
                amount(),
                expenseCategory);
    }

    /**
     * Represents a unique identifier for a transaction.
     * This identifier is used to distinguish transactions within the system.
     * It is immutable and encapsulates a single string value.
     * <p/>
     * Inline classes, also known as value classes, provide a way to create immutable and memory-efficient data structures.
     * This will improve performance and reduce memory overhead by avoiding object allocation and enabling better
     * optimizations by the JVM. When <a href="https://openjdk.org/projects/valhalla/">Project Valhalla</a> is released, consider refactoring this class to take advantage
     * of these benefits.
     */
    public record TransactionId(String value) {
        public TransactionId {
            Objects.requireNonNull(value, "Transaction ID cannot be null");
            if (value.isBlank()) {
                throw new IllegalArgumentException("Transaction ID cannot be null or blank");
            }
        }

        // concise modern factory
        public static TransactionId of(String id) {
            return new TransactionId(id);
        }
    }

    public record ClientId(String value) {
        public ClientId {
            Objects.requireNonNull(value, "Client ID cannot be null");
            if (value.isBlank()) {
                throw new IllegalArgumentException("Client ID cannot be null or blank");
            }
        }

        // concise modern factory
        public static ClientId of(String id) {
            return new ClientId(id);
        }
    }

    public record AccountId(String value) {
        public AccountId {
            Objects.requireNonNull(value, "Account ID cannot be null");
            if (value.isBlank()) {
                throw new IllegalArgumentException("Account ID cannot be null or blank");
            }
        }

        // concise factory
        public static AccountId of(String id) {
            return new AccountId(id);
        }
    }

    public record Amount(BigDecimal value) {
        public Amount {
            Objects.requireNonNull(value, "Amount cannot be null");
        }

        // concise factory
        public static Amount of(BigDecimal value) {
            return new Amount(value);
        }
    }

    public record MerchantCategoryCode(String value) {
        public MerchantCategoryCode {
            Objects.requireNonNull(value, "Merchant category code cannot be null");
            if (!value.matches("\\d{4}")) {
                throw new IllegalArgumentException("Merchant category code must be exactly 4 digits");
            }
        }

        // concise factory
        public static MerchantCategoryCode of(String value) {
            return new MerchantCategoryCode(value);
        }
    }
}

File: fvf4k/src/main/kotlin/fvf4k/demo/domain/model/Transaction.kt
package fvf4k.demo.domain.model

import arrow.core.raise.catch
import arrow.core.raise.context.Raise
import arrow.core.raise.context.raise
import fvf4k.demo.domain.failure.InvalidCurrencyCode
import fvf4k.demo.domain.failure.InvalidMerchantCategoryPattern
import fvf4k.demo.domain.failure.NullMerchantCategoryCode
import fvf4k.demo.domain.failure.NullOrEmpty
import fvf4k.demo.domain.failure.ValidationFailed
import java.math.BigDecimal
import java.util.*


data class Transaction(
    val id: TransactionId,
    val clientId: ClientId,
    val accountId: AccountId,
    val money: Money,
    val mcc: MerchantCategoryCode
)

@JvmInline value class ClientId private constructor(val value: UUID) {
    companion object {
        context(_: Raise<ValidationFailed>)
        operator fun invoke(value: String?): ClientId {
            if (value == null || value.isEmpty()) raise(NullOrEmpty("clientId", value))
            else return ClientId(value)
        }

        context(_: Raise<ValidationFailed>)
        operator fun invoke(value: UUID?): ClientId {
            if (value == null) raise(NullOrEmpty("clientId", value))
            else return ClientId(value)
        }
    }
}

@JvmInline value class AccountId private constructor(val value: UUID) {
    companion object {
        context(_: Raise<ValidationFailed>)
        operator fun invoke(value: String?): AccountId {
            if (value == null || value.isEmpty()) raise(NullOrEmpty("accountId", value))
            else return AccountId(value)
        }

        context(_: Raise<ValidationFailed>)
        operator fun invoke(value: UUID?): AccountId {
            if (value == null) raise(NullOrEmpty("accountId", value))
            else return AccountId(value)
        }
    }
}

data class Money private constructor(val value: BigDecimal, val currency: Currency) {
    companion object {
        context(_: Raise<ValidationFailed>)
        operator fun invoke(value: BigDecimal?, currencyCode: String?): Money {
            if (currencyCode == null || currencyCode.isEmpty()) raise(NullOrEmpty("currencyCode", currencyCode))
            val currency = catch({ Currency.getInstance(currencyCode) }) { raise(InvalidCurrencyCode(currencyCode)) }
            if (value == null) raise(NullOrEmpty("amount", value))
            return Money(value, currency)
        }
    }
}

@JvmInline value class TransactionId(val value: UUID) {
    companion object {
        context(_: Raise<ValidationFailed>)
        operator fun invoke(value: String?): TransactionId {
            if (value == null || value.isEmpty()) raise(NullOrEmpty("transactionId", value))
            else return TransactionId(value)
        }

        context(_: Raise<ValidationFailed>)
        operator fun invoke(value: UUID?): TransactionId {
            if (value == null) raise(NullOrEmpty("transactionId", value))
            else return TransactionId(value)
        }
    }
}

@JvmInline value class MerchantCategoryCode private constructor(val value: String) {
    companion object {
        private val MCC_PATTERN = Regex("^\\d{4}$")

        context(_: Raise<ValidationFailed>)
        operator fun invoke(value: String?): MerchantCategoryCode = when {
            value == null -> raise(NullMerchantCategoryCode)
            !MCC_PATTERN.matches(value) -> raise(InvalidMerchantCategoryPattern(value))
            else -> MerchantCategoryCode(value)
        }
    }
}
